openshift-kube-apiserver/authorization/minimumkubeletversion/minimum_kubelet_version.go

```diff
@@ -0,0 +1,86 @@
+package minimumkubeletversion
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/blang/semver/v4"
+	nodelib "github.com/openshift/library-go/pkg/config/node"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	corev1informers "k8s.io/client-go/informers/core/v1"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/auth/nodeidentifier"
+)
+
+var minimumKubeletVersion *semver.Version
```

Comment:
Accidentally dropping the call to SetMinimumKubeletVersion during a rebase seems like a mistake that could realistically be made. Could you clamp down on this variable so that we panic if it is assigned twice or read before assignment, and that those panics still work if there are concurrent reads and writes?

Reply:
read before assignment is okay I think, we need to handle the case it's not set. The one way I can think of doing that is unconditinoally call SetMinimumKubeletVersion and interpret "" as KAS-3 version, but that'd require some plumbing and I'm not sure it's worth it. I think proper e2e tests would protect against a rebase breaking the feature.  I can fix the double set case

Reply:
What is meant to happen if no minimum kubelet version is configured? Right now it looks like IsNodeTooOld returns an error and the authorizer will always deny. The config field is optional, I would have expected the authorizer to make no decision if unconfigured.

Maybe it'd be clearer to slap a `minimumKubeletVersion` field onto the `minimumKubeletVersionAuth` struct so that you can unit test it without mutating shared global state, then access your global state inside `NewMinimumKubeletVersion`. If you tweak the signature to return `authorizer.Authorizer`, then it could return a static "no decision" authorizer when no minimum version is being enforced, and a `*minimumKubeletVersionAuth` otherwise.

Mutable global state is easy to break, so I suggest a sync.Once that is used to coordinate a panic for both "assigned twice" and "assigned after read", plus a clear explanation in the comments explaining how the global state is expected to be initialized (exactly once or at most once, and at what time relative to first read).

Reply:
is there a different way I can say NoDecision: https://github.com/openshift/kubernetes/pull/2104/files#diff-b7bdcb41a84b0a21cc83959481df9b457492b2e10b6d1350dff1278c8489bf44R45 ? This looks to me like it will defer to other plugins to approve, but i've seen comments in other places around NoDecisions that imply NoDecision is a denial